File: app/src/main/java/com/commit451/gitlab/adapter/GroupMembersAdapter.java
package com.commit451.gitlab.adapter;

import android.support.v7.widget.RecyclerView;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.PopupMenu;

import com.commit451.gitlab.R;
import com.commit451.gitlab.model.api.Member;
import com.commit451.gitlab.viewHolder.LoadingFooterViewHolder;
import com.commit451.gitlab.viewHolder.ProjectMemberViewHolder;

import java.util.ArrayList;
import java.util.Collection;

/**
 * Adapter for a list of users
 */
public class GroupMembersAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder>  {

    private static final int TYPE_MEMBER = 0;
    private static final int TYPE_FOOTER = 1;

    private static final int FOOTER_COUNT = 1;

    private Listener listener;
    private ArrayList<Member> values;
    private boolean loading = false;

    public GroupMembersAdapter(Listener listener) {
        this.listener = listener;
        values = new ArrayList<>();
    }

    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        switch (viewType) {
            case TYPE_MEMBER:
                ProjectMemberViewHolder holder = ProjectMemberViewHolder.inflate(parent);
                holder.itemView.setOnClickListener(new View.OnClickListener() {
                    @Override
                    public void onClick(View v) {
                        int position = (int) v.getTag(R.id.list_position);
                        ProjectMemberViewHolder holder = (ProjectMemberViewHolder) v.getTag(R.id.list_view_holder);
                        listener.onUserClicked(getMember(position), holder);
                    }
                });
                return holder;
            case TYPE_FOOTER:
                return LoadingFooterViewHolder.inflate(parent);
        }
        throw new IllegalStateException("No known ViewHolder for type " + viewType);

    }

    @Override
    public void onBindViewHolder(RecyclerView.ViewHolder holder, int position) {
        if (holder instanceof ProjectMemberViewHolder) {
            final Member member = values.get(position);
            ((ProjectMemberViewHolder) holder).bind(member);
            ((ProjectMemberViewHolder) holder).popupMenu.setOnMenuItemClickListener(new PopupMenu.OnMenuItemClickListener() {
                @Override
                public boolean onMenuItemClick(MenuItem item) {
                    switch (item.getItemId()) {
                        case R.id.action_change_access:
                            listener.onUserChangeAccessClicked(member);
                            return true;
                        case R.id.action_remove:
                            listener.onUserRemoveClicked(member);
                            return true;
                    }
                    return false;
                }
            });
            holder.itemView.setTag(R.id.list_position, position);
            holder.itemView.setTag(R.id.list_view_holder, holder);
        } else if(holder instanceof LoadingFooterViewHolder) {
            ((LoadingFooterViewHolder) holder).bind(loading);
        }
    }

    @Override
    public int getItemViewType(int position) {
        if (position == values.size()) {
            return TYPE_FOOTER;
        } else {
            return TYPE_MEMBER;
        }
    }

    @Override
    public int getItemCount() {
        return values.size() + FOOTER_COUNT;
    }

    private Member getMember(int position) {
        return values.get(position);
    }

    public void setData(Collection<Member> members) {
        values.clear();
        addData(members);
    }

    public void addData(Collection<Member> members) {
        if (members != null) {
            values.addAll(members);
        }
        notifyDataSetChanged();
    }

    public void addMember(Member member) {
        values.add(0, member);
        notifyItemInserted(0);
    }

    public void removeMember(Member member) {
        int index = values.indexOf(member);
        values.remove(index);
        notifyItemRemoved(index);
    }

    public void setLoading(boolean loading) {
        this.loading = loading;
        notifyItemChanged(values.size());
    }

    public boolean isLoading() {
        return loading;
    }

    public boolean isFooter(int position) {
        int viewType = getItemViewType(position);
        if (viewType == TYPE_FOOTER) {
            return true;
        }
        return false;
    }

    public interface Listener {
        void onUserClicked(Member member, ProjectMemberViewHolder userViewHolder);
        void onUserRemoveClicked(Member member);
        void onUserChangeAccessClicked(Member member);
    }
}

File: app/src/main/java/com/commit451/gitlab/adapter/GroupMembersAdapter.kt
package com.commit451.gitlab.adapter

import android.support.v7.widget.RecyclerView
import android.view.ViewGroup
import android.widget.PopupMenu
import com.commit451.gitlab.R
import com.commit451.gitlab.model.api.User
import com.commit451.gitlab.viewHolder.LoadingFooterViewHolder
import com.commit451.gitlab.viewHolder.ProjectMemberViewHolder
import java.util.*

/**
 * Adapter for a list of users
 */
class GroupMembersAdapter(private val listener: GroupMembersAdapter.Listener) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {

    companion object {

        private val TYPE_MEMBER = 0
        private val TYPE_FOOTER = 1

        private val FOOTER_COUNT = 1
    }

    val values: ArrayList<User> = ArrayList()

    var isLoading = false
        set(loading) {
            field = loading
            notifyItemChanged(values.size)
        }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        when (viewType) {
            TYPE_MEMBER -> {
                val holder = ProjectMemberViewHolder.inflate(parent)
                holder.itemView.setOnClickListener { v ->
                    val position = v.getTag(R.id.list_position) as Int
                    listener.onUserClicked(getMember(position), holder)
                }
                return holder
            }
            TYPE_FOOTER -> return LoadingFooterViewHolder.inflate(parent)
        }
        throw IllegalStateException("No known ViewHolder for type " + viewType)

    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        if (holder is ProjectMemberViewHolder) {
            val member = values[position]
            holder.bind(member)
            holder.popupMenu.setOnMenuItemClickListener(PopupMenu.OnMenuItemClickListener { item ->
                when (item.itemId) {
                    R.id.action_change_access -> {
                        listener.onUserChangeAccessClicked(member)
                        return@OnMenuItemClickListener true
                    }
                    R.id.action_remove -> {
                        listener.onUserRemoveClicked(member)
                        return@OnMenuItemClickListener true
                    }
                }
                false
            })
            holder.itemView.setTag(R.id.list_position, position)
            holder.itemView.setTag(R.id.list_view_holder, holder)
        } else if (holder is LoadingFooterViewHolder) {
            holder.bind(isLoading)
        }
    }

    override fun getItemViewType(position: Int): Int {
        if (position == values.size) {
            return TYPE_FOOTER
        } else {
            return TYPE_MEMBER
        }
    }

    override fun getItemCount(): Int {
        return values.size + FOOTER_COUNT
    }

    fun setData(members: Collection<User>?) {
        values.clear()
        addData(members)
    }

    fun addData(members: Collection<User>?) {
        if (members != null) {
            values.addAll(members)
        }
        notifyDataSetChanged()
    }

    fun addMember(member: User) {
        values.add(0, member)
        notifyItemInserted(0)
    }

    fun removeMember(member: User) {
        val index = values.indexOf(member)
        values.removeAt(index)
        notifyItemRemoved(index)
    }

    fun isFooter(position: Int): Boolean {
        val viewType = getItemViewType(position)
        if (viewType == TYPE_FOOTER) {
            return true
        }
        return false
    }

    private fun getMember(position: Int): User {
        return values[position]
    }

    interface Listener {
        fun onUserClicked(member: User, userViewHolder: ProjectMemberViewHolder)
        fun onUserRemoveClicked(member: User)
        fun onUserChangeAccessClicked(member: User)
    }
}
